### ISSUE TITLE
Dry run mode for transaction submission

### ISSUE THREAD

**Issue by bartekn** [stellar/stellar-core#1391]
# Issue type
  feature request

# Issue description
Would be great to have a _dry run_ mode for transaction submission. It would check if transaction is valid and if would make it into ledger but doesn't actually submit it to the network. It doesn't check signatures.

Possible use cases:
* Multisig transactions that require lengthy process of signing.
* Smart contracts.

**Comment by MonsieurNicolas** [comment]
Hey - thanks for the suggestion. I am not sure I understand how this helps for either though?

Are you trying to:

A) see if a transaction would be accepted by core if it was submitted? The checks that we perform at submission time are quite limited (things like checking for sequence number), that I am not sure having a different operating mode would help much for either of those scenarios as the complexity is typically baked in the interaction of transactions, which would require B)
Or
B) Are you trying to check if a transaction would succeed if it was submitted and processed by the network? That is quite expensive (and also doesn't give a guarantee that it will succeed later) and right now we want people to pay (through fees) to mitigate for this.

**Comment by bartekn** [comment]
Example from some time ago: 3 signers signed a transaction with payment operation, once submitted it turned out account doesn't exist (create_account was required).

I think checking for such things like below (and more) would be helpful:
* Sequence number (is correct)
* Accounts existence (create_account vs payment, does tx/op source exists, for account_merge does account exists?)
* Balance (is sending more than can)
* Change_trust - is current balance enough for subentries?
* For more complicated things: for allow_trust is auth_required set?

I know there are limits and state can change between test and actual submission but I think something like this could be helpful.

**Comment by MonsieurNicolas** [comment]
yeah so it sounds like this is more like what the "B" endpoint would do more or less... and then people will want to run a sequence of transactions through this system, and we'll be back to the same place. Is it that it's hard for people to test their code somehow? Do we need to provide a better environment than what testnet does today and/or provide some endpoints to improve the test coverage?

**Comment by mvaneijk** [comment]
Wouldn't it be better to add it to Horizon? All transaction resource information is available in that database, and Horizon look-ups are "free".

**Comment by bartekn** [comment]
I think we should have some way for testing transactions/smart contracts. The problem with testnet is that hashes and sequence numbers will differ between networks. Consider something like vesting accounts (locking funds for a certain amount of time). This can be checked in testnet but it's still possible to make a mistake when recreating the same configuration in live network (typos, invalid sequence, mistakes when changing time bounds etc). It would be great to have a testing tool that will check if transaction(s) will succeed given the current state of the set of accounts in the network.

My high level vision for this tool is something like:
1. Create a local fork of the current state of the network locally (let's call the fork ledger `A`). Now, only tester can submit transactions.
2. Tests scenarios always start from ledger `A` (when first test ends, we're back at ledger `A` and second test is started and so on...).
3. Tester should be able to change the timestamp of the next ledger to a timestamp in a future (so we can test time bounds).

@mvaneijk I think this should be tied to core somehow because rewriting core code somewhere else will certainly introduce incompatibilities.

**Comment by MonsieurNicolas** [comment]
yeah I think @bartekn is probably right on the type of tool that we would build. Basically what we're talking about here is some sort of debugger/simulator which seems fine.
I do see the value during the coding phase of a project, I am quite skeptical about using it in the normal flow for submitting transactions which is why I think there is a bit of thinking to do on the semantics with the network and how we want people to write tests so that they have high confidence in their code.

**Comment by bartekn** [comment]
We discussed this again last week. Alternatively, we can create a docker image that starts with one argument: tx envelope to test. Then:
1. It syncs to the latest ledger in a given network.
2. Changes the config of stellar-core to standalone network (so that tx is not submitted to the rest of the network during tests).
3. Start stellar-core with changed config file and submits a transaction.
4. Returns results, meta, etc. to the user.

CC: @tomquisel
